src/lex/cpp.rs

```diff
@@ -481,15 +703,130 @@ impl<'a> PreProcessor<'a> {
+                    data: Token::Id(id),
+                    location,
+                })) => (id, location),
+                Some(Err(err)) => return Err(err),
+                Some(Ok(other)) => {
+                    return Err(CompileError::new(
+                        CppError::UnexpectedToken("include file", other.data).into(),
+                        other.location,
+                    ))
+                }
+                None => {
+                    return Err(CompileError::new(
+                        CppError::EndOfFile("include file").into(),
+                        self.span(start),
+                    ))
+                }
+            };
+            match self.replace_id(id, location) {
+                // local
+                Some(Ok(Locatable {
+                    data: Token::Literal(Literal::Str(_)),
+                    ..
+                })) => unimplemented!("#include for macros"), //return self.include_path(id, true, start),
+                // system
+                Some(Ok(Locatable {
+                    data: Token::Comparison(ComparisonToken::Less),
+                    ..
+                })) => false,
+                Some(Ok(other)) => {
+                    return Err(CompileError::new(
+                        CppError::UnexpectedToken("include file", other.data).into(),
+                        other.location,
+                    ))
+                }
+                Some(Err(err)) => return Err(err),
+                None => {
+                    return Err(CompileError::new(
+                        CppError::EndOfFile("include file").into(),
+                        self.span(start),
+                    ))
+                }
+            }
+        };
+
+        let filename = self.bytes_until(if local { b'"' } else { b'>' });
+        self.include_path(filename, local, start)
+    }
+    // we've done the parsing for an `#include`,
+    // now we want to do the dirty work of reading it into memory
+    fn include_path(
+        &mut self,
+        filename: Vec<u8>,
+        local: bool,
+        start: u32,
+    ) -> Result<(), Locatable<Error>> {
+        const SEARCH_PATH: &[&str] = &["/usr/include"];
+        log::debug!("in search path");
+
+        // Recall that the original file was valid UTF8.
+        // Since in UTF8 no ASCII character can occur
+        // within a multi-byte sequence, `filename` must be valid UTF8.
+        let filename = String::from_utf8(filename).expect("passed invalid utf8 to start");
+
+        // local include: #include "dict.h"
+        if local {
+            // TODO: relative file paths
+            unimplemented!();
+        }
+        // if we don't find it locally, we fall back to system headers
+        // this is part of the spec!
+        for path in SEARCH_PATH {
+            let mut buf = PathBuf::from(path);
+            buf.push(&filename);
+            if buf.exists() {
+                // TODO: _any_ sort of error handling
```

Comment:
Unfortunately `io::Error` implements neither `PartialEq` nor `Clone`, so it's not that simple. Making a new `FatalError` enum doesn't help because we still couldn't put it in `Error`. I'm open to ideas.

Reply:
For now, I just called `.to_string()` on the IO Error.